add Python getting started guide (#55)
memgraph/documentation: pages/client-libraries/python.mdx

    # Find a user John in the database
    records, summary, keys = driver.execute_query(
        "MATCH (u:User {name: $name}) RETURN u.name AS name",
        name="John",
        database_="memgraph",
    )

    # Get the result
    for record in records:
        print(record["name"])

    # Print the query
    print(summary.query)
```

### Run the application

Now, you can run the application with the following command:

```
python3 main.py
```

If everything was set up correctly, you should see the following output in the terminal: 

```
John
{'labels_added': 1, 'labels_removed': 0, 'nodes_created': 1, 'nodes_deleted': 0, 'properties_set': 0, 'relationships_created': 0, 'relationships_deleted': 0}
John
MATCH (u:User {name: $name}) RETURN u.name AS name
```

### Visualize data

To visualize objects created in the database with the `main.py` script, head over to [http://localhost:3000/](http://localhost:3000/) and run `MATCH (n) RETURN n;` in the *Query Execution* tab. 
That query will visualize the created node. By clicking on a node, you can explore different properties.

![python-quick-start](/pages/client-libraries/new-python/python-quick-start.png)

### Next steps 

> This makes a good starting point for Python application. For more information on how to use the Python driver, 
> continue reading the guide under [Python client API usage and examples](#python-client-api-usage-and-examples). 

Reviewer: ```suggestion
You can continue building your Python application. For more information on how to use the Python driver, 
continue reading about [Python client API usage and examples](#python-client-api-usage-and-examples). 
```

</Steps>

## Python client API usage and examples

After a brief [Quickstart guide](#quickstart), this section will go into more detail on how to use the Python driver API, explain code snippets, and provide more examples. Feel free to skip to the section that interests you the most.

### Database connection 

> Once the database is running and the driver is installed or available in Python, you should be able to connect to the database in the following ways:

Reviewer: ```suggestion
Once the database is running and the driver is installed or available in Python, you should be able to connect to the database in one of the following ways:
```

    # Count the number of nodes in the database
    records, summary, keys = driver.execute_query(
        "CREATE (u:User {name: $name, password: $password}) RETURN u.name AS name",
        name="John",
        password="pass",
        database_="memgraph",
    )

    # Get the result
    for record in records:
        print(record["name"])
```

### Query the database 

After connecting your driver to Memgraph, you can start running queries. The simplest way to run queries is by using the `execute_query()` method which has an automatic transaction management. 

#### Run a create query 

The following query will create a node inside the database: 

```python
node = "CREATE (n:Technology {name:'Memgraph'});"
driver.execute_query(node)
```

Due to the nature of the `execute_query()` method, transactions are handled automatically. 

#### Run a read query 

The following query will read the previously created node from the database: 

```python
node = "MATCH (n:Technology) RETURN n;"
result = driver.execute_query(node)

for record in records:
    node = record["n"] 
    print(node)   

```

In this query, each record contains a node object behind the variable `n`.

> 

Reviewer: ```suggestion
```

#### Running queries with property map

```python
query = "CREATE (n:Technology {id:$id, name: $name, description: $description});"
driver.execute_query(query, 
    id=1, 
    name="MemgraphDB", 
    description="Fastest graph database")
```

Using this approach, the queries will not contain hard-coded values, they can be more dynamic.

### Proces the results 

> Processing results from the database is important since we do not want to lose any data during conversions and properly read results and serve them back to the Python application. Python is dynamically typed language, which means that the type of the variable is determined at runtime, which is why we need to be careful when processing results.

Reviewer: ```suggestion
Processing results from the database is important to avoid losing any data during conversions and properly read results and serve them back to the Python application. Python is a dynamically typed language, which means that the type of the variable is determined at runtime, which is why we need to be careful when processing results.
```

#### Process the Node result

> In order to process the result you need to read them first, you can do that by running the following query:

Reviewer: ```suggestion
To process the results, you need to read them first. Do that by running the following query:
```

```python
query = "MATCH (n:Technology) RETURN n;"
records, summary, keys = driver.execute_query(query)
```
The `execute_query()` method returns a tuple of three elements: `records`, `summary` and `keys`. The records field contains all the records returned by the query. To process the results, you can iterate over the records and access the fields you need. 

For example: 
    
```python
for record in records:
    node = record["n"]
    print(node)
```
```text
<Node element_id='48' labels=frozenset({'Technology'}) properties={'description': 'Fastest graph database', 'id': 1, 'name': 'MemgraphDB'}>
```

> In the example above, each returned record, you can access the `n` field, which is a `Node` returned from a query.

Reviewer: This is a very strange sentence

Reviewer 2: ```suggestion
As you can see in the example above, each returned record has a `Node` returned from a query.
```

```python
print(node)             # <Node element_id='47' labels=frozenset({'Technology'}) properties={'description': 'Fastest graph database', 'id': 1, 'name': 'MemgraphDB'}>
print(node.labels)      # frozenset({'Technology'})
print(node.element_id)  # 48
print(node["id"])       # 1
print(node["name"])     # MemgraphDB
```

Keep in mind that the `element_id` property returns the internal ID of the node, which is not the same as the user-defined ID, and it should not 
be used for any application-level logic.

#### Process the Relationship result

You can also receive a relationship from a query. For example: 

```python
    query = "CREATE (u:User {name: 'John'})-[r:LOVES {id:99}]->(t:Technology {name: 'Memgraph'})"
    driver.execute_query(query)

    query = "MATCH (u:User)-[r:LOVES]->(t:Technology) RETURN r"
    records, summary, keys = driver.execute_query(query)
    for record in records:
        relationship = record["r"]
        print(relationship)
        print(relationship.element_id)
        print(relationship["id"])
        print(relationship.type)
        print(relationship.nodes)

```

> You can access the `Relationship` properties just like the [Node](#process-the-node-result) properties. Keep in mind that `element_id` is the internal ID of the relationship, which is not the same as the user-defined ID, and it should not be used for any application-level logic.

Reviewer: ```suggestion

You can access the `Relationship` properties just like the [Node](#process-the-node-result) properties. Keep in mind that `element_id` is the internal ID of the relationship, which is not the same as the user-defined ID, and it should not be used for any application-level logic.
```

If you want more control over the query execution, use managed transactions to run the queries.

To run a managed transaction, you first need to create a **session**:

```python
my_session = driver.session(database="memgraph")
# do something with the session
my_session.close()
```

Don't forget to close the session after usage with `session.close()` or Python `with` statement, which will automatically close a session, like in the following code snippet:

```python
with driver.session(database="memgraph") as session:
    # do something with the session

# session is automatically closed
```

Sessions are not thread safe, so make sure that each thread creates its own sessions. On the other hand, the main driver object can be shared across threads.
One transaction can contain multiple queries, and they will either all be executed or none will. This means that you don't have to worry about rolling back the part of the changes that got executed in that transaction and finding the ones that didn't.
Having more than one query in a transaction is useful when the queries work on a similar database task, usually creating graph database objects.

> With sessions, you can run:

Reviewer: ```suggestion
With sessions, you can:
```

To query the database, run the `execute_query()` procedure on the driver object with the Cypher query argument, along with its parameters and database name.
A good practice is to always provide a database name, to avoid extra request to the server and in that way, improve performance and to provide parameters to protect your queries from Cypher injections.
Since v2.10, Memgraph added the [multi-tenant support](/configuration/multi-tenancy) in the Enterprise Edition to manage multiple isolated databases within a single instance.
Make sure you provide the correct database name if you're working with more than one tenant.

The `execute_query()` procedure automatically creates a transaction in which you can include multiple Cypher statements in a single query. If a transaction fails, it automatically retries to run the transaction.

Bolt protocol specifies additional data that can be sent along with the requested results. Such data is called [metadata](/configuration/metadata) and can be divided into two groups: query statistics and notifications.
From the query statistics metadata we can get query counters that tell us which changes the **write query** triggered on the server.

> ```python

Reviewer 2: Yes

To run a transaction, you first need to create a **session**:

```python
my_session = driver.session(database="memgraph")
# do something with the session
my_session.close()
```

Don't forget to close the session after usage with `session.close()` or use `with` statement in Python which will automatically close a session, like in the following code snippet:

```python
with driver.session(database="memgraph") as session:
    # do something with the session

# session is automatically closed
```

Sessions are not thread safe, so make sure that each thread creates its own sessions. On the other hand, the main driver object can be shared across threads.
One transaction can contain multiple queries and they will either all be executed or none of them. This means that you don't have to worry about rolling back the part of the changes that got executed in that transaction and finding the ones that didn't.
Having more than one query in a transaction is useful when those queries work on a similar database task, usually creating graph database objects.

With sessions, you can run:

> - [Manged transactions](#managed-transactions) - to run multiple queries with automatic retries without the possibility to roll back a query within a transaction.

Reviewer 2: Retry is better. Rerun associates me with the whole process of rerunning, while retry is the procedure or incremental

    with GraphDatabase.driver(URI, auth=AUTH) as driver:
        create_users(driver, sender, receiver)
        transfer_tokens(driver, 1, 2, 10)

if __name__ == "__main__":
    main()
```

In the above example, if you change John's account balance to a number less than 10, you will see the warning that he doesn't have enough tokens and the transfer won't happen (it will be rolled back).

##### Implicit transactions

Implicit or auto-commit transactions are the simplest way to run a Cypher query since they won't be automatically retried as with `execute_query()` procedure or managed transactions.
With implicit transactions, you don't have the same control of transaction as with explicit transactions, so they are mostly used for quick prototyping.

To run an implicit transaction, use `Session.run()` method:

```python
with GraphDatabase.driver(URI, auth=AUTH) as driver:
    with driver.session(database="memgraph") as session:
        session.run("CREATE (u:User {name: $name})", name="John")
```

The `Session.run()` method is most commonly used for `LOAD CSV` clause to prevent timeout errors due to the size of the transaction.

#### Concurrent transactions

> It is possible to run concurrent transactions with Python's `asyncio` library through `AsyncGraphDatabase` driver.

Reviewer 2: ohh boy 😅 😅

> The code is not that different from the synchronous version, except that you must use await on all async calls, and define as async all functions that should be awaited.

Reviewer 2: It is a keyword that should be understandable without extra context.

Here is how you can use the driver with the `execute_query()` method:

```python
import asyncio
from neo4j import AsyncGraphDatabase

URI = "bolt://localhost:7687"
AUTH = ("", "")

async def main():
    async with AsyncGraphDatabase.driver(URI, auth=AUTH) as driver:
        records, summary, keys = await driver.execute_query(
            "MATCH (u:User) RETURN u.name AS name", database_="memgraph"
        )
        names = [record["name"] for record in records]
        print(names)

if __name__ == "__main__":
    asyncio.run(main())
```

Here is how you can use the driver with transaction functions:

```python
import asyncio
from neo4j import AsyncGraphDatabase

URI = "bolt://localhost:7687"
AUTH = ("", "")

async def main():
    async with AsyncGraphDatabase.driver(URI, auth=AUTH) as driver:
        async with driver.session(database="memgraph") as session:
            records = await session.execute_read(get_users)
            print(records)

async def get_users(tx):
    result = await tx.run("MATCH (u:User) RETURN u.name AS name")
    records = await result.values()
    return records

if __name__ == "__main__":
    asyncio.run(main())
```

> ## Where to next?

Reviewer 2: Deleted
